R/utils.R

```diff
@@ -122,76 +122,124 @@ report_card_template <- function(title, label, description = NULL, with_filter,
+      paste(modules_msg, collapse = "\n"),
+      toString(dQuote(datanames, q = FALSE))
+    )
+  } else {
+    TRUE
+  }
+}
+
+#' @rdname check_modules_datanames
+check_modules_datanames_html <- function(modules,
+                                         datanames) {
+  check_datanames <- check_modules_datanames_recursive(modules, datanames)
+  show_module_info <- inherits(modules, "teal_modules")
+  if (!length(check_datanames)) {
+    return(TRUE)
+  }
+  shiny::tagList(
+    lapply(
+      check_datanames,
+      function(mod) {
+        tagList(
+          tags$span(
+            tags$span(`if`(length(mod$missing_datanames) > 1, "Datasets", "Dataset")),
+            to_html_code_list(mod$missing_datanames),
+            tags$span(
+              paste0(
+                `if`(length(mod$missing_datanames) > 1, "are missing", "is missing"),
+                `if`(show_module_info, sprintf(" for tab '%s'.", mod$label), ".")
+              )
             ),
-            tagList = function(...) trimws(paste(...))
+            if (length(mod$datanames) >= 1) {
+              tagList(
+                tags$span(`if`(length(mod$datanames) > 1, "Datasets", "Dataset")),
+                tags$span("available in data:"),
+                tagList(
+                  tags$span(
+                    to_html_code_list(mod$datanames),
+                    tags$span(".", .noWS = "outside"),
+                    .noWS = c("outside")
+                  )
+                )
+              )
+            } else {
+              tags$span("No datasets are available in data.")
+            }
           ),
-          # Build HTML representation of the error message with <pre> formatting
-          html = function(with_module_name = TRUE) {
-            tagList(
-              build_datanames_error_message(
-                if (with_module_name) modules$label,
-                datanames,
-                extra_datanames
-              ),
-              tags$br(.noWS = "before")
-            )
-          }
+          tags$br(.noWS = "before")
         )
       }
-    }
-  }
-  check_datanames <- recursive_check_datanames(modules, datanames)
-  if (length(check_datanames)) {
-    check_datanames
+    )
+  )
+}
+
+#' Recursively checks modules and returns list for every datanames mismatch between module and data
+#' @noRd
+check_modules_datanames_recursive <- function(modules, datanames) { # nolint: object_name_length
+  checkmate::assert_multi_class(modules, c("teal_module", "teal_modules"))
+  checkmate::assert_character(datanames)
+  if (inherits(modules, "teal_modules")) {
+    unlist(
+      lapply(modules$children, check_modules_datanames_recursive, datanames = datanames),
+      recursive = FALSE
+    )
   } else {
-    TRUE
+    missing_datanames <- setdiff(modules$datanames, c("all", datanames))
+    if (length(missing_datanames)) {
+      list(list(
+        label = modules$label,
+        dataname = modules$datanames,
```

Comment:
Typo, I believe this element is never used though. (see other comment)

```suggestion
        datanames = modules$datanames,
```